src/Batch/CsvImportBatch.php

```diff
@@ -0,0 +1,117 @@
+<?php
+
+namespace Drupal\csvimport\Batch;
+
+// @codingStandardsIgnoreStart
+use Drupal\node\Entity\Node;
+
+// @codingStandardsIgnoreEnd
+
+/**
+ * Methods for running the CSV import in a batch.
+ *
+ * @package Drupal\csvimport
+ */
+class CsvImportBatch {
+
+  /**
+   * Handle batch completion.
+   *
+   *   Creates a new CSV file containing all failed rows if any.
+   */
+  public static function csvimportImportFinished($success, $results, $operations) {
+
+    $messenger = \Drupal::messenger();
+
+    if (!empty($results['failed_rows'])) {
+
+      $dir = 'public://csvimport';
+      if (file_prepare_directory($dir, FILE_CREATE_DIRECTORY)) {
+
+        // We validated extension on upload.
+        $csv_filename = 'failed_rows-' . basename($results['uploaded_filename']);
+        $csv_filepath = $dir . '/' . $csv_filename;
+
+        $targs = [
+          ':csv_url'      => file_create_url($csv_filepath),
+          '@csv_filename' => $csv_filename,
+          '@csv_filepath' => $csv_filepath,
+        ];
+
+        if ($handle = fopen($csv_filepath, 'w+')) {
+
+          foreach ($results['failed_rows'] as $failed_row) {
+            fputcsv($handle, $failed_row);
+          }
+
+          fclose($handle);
+          $messenger->addMessage(t('Some rows failed to import. You may download a CSV of these rows: <a href=":csv_url">@csv_filename</a>', $targs), 'error');
+        }
+        else {
+          $messenger->addMessage(t('Some rows failed to import, but unable to write error CSV to @csv_filepath', $targs), 'error');
+        }
+      }
+      else {
+        $messenger->addMessage(t('Some rows failed to import, but unable to create directory for error CSV at @csv_directory', $targs), 'error');
+      }
+    }
+
+    return t('The CSV import has completed.');
+  }
+
+  /**
+   * Remember the uploaded CSV filename.
+   *
+   * @TODO Is there a better way to pass a value from inception of the batch to
+   * the finished function?
+   */
+  public static function csvimportRememberFilename($filename, &$context) {
+
+    $context['results']['uploaded_filename'] = $filename;
+  }
+
+  /**
+   * Process a single line.
+   */
+  public static function csvimportImportLine($line, &$context) {
+
+    $context['results']['rows_imported']++;
+    $line = array_map('base64_decode', $line);
+
+    // Simply show the import row count.
+    $context['message'] = t('Importing row !c', ['!c' => $context['results']['rows_imported']]);
+
+    // Alternatively, our example CSV happens to have the title in the
+    // third column, so we can uncomment this line to display "Importing
+    // Blahblah" as each row is parsed.
+    //
+    // You can comment out the line above if you uncomment this one.
+    $context['message'] = t('Importing %title', ['%title' => $line[2]]);
+
+    // In order to slow importing and debug better, we can uncomment
+    // this line to make each import slightly slower.
+    usleep(2500);
```

Comment:
I feel like this should be commented by default, both because otherwise some folks will not spot it (:exclamation: :exclamation:) and because it should match the comment above. Not sure yet if this change was introduced in this MR!

Reply:
Good catch. I think I'm going to refresh all the documentation on the next PR as well.